2021/17.py

```diff
@@ -3,22 +3,31 @@
 import heapq
 import itertools
 from collections import defaultdict, Counter, deque
+from math import ceil, sqrt
+
+#target area: x=20..30, y=-10..-5
+# MinX, MaxX = (20, 30)
+# MinY, MaxY = (-10, -5)
 
 #target area: x=96..125, y=-144..-98
+MinX, MaxX = (96, 125)
+MinY, MaxY = (-144, -98)
 
 # y(t) ~ -t^2 + DY*t
 
 p2 = 0
 ans = 0
-for DX in range(150):
-    for DY in range(-150, 1000):
+
+minDX = int(ceil((sqrt(8*MinX+1)-1)/2))
+for DX in range(minDX, MaxX+1):
+    for DY in range(MinY, 2-MinY):
```

Comment:
The min `DY` is also obvious: the probe steps immediately to the end of the target range.

The max `DY` is a bit less obvious: what goes up must come down and after the probe reaches its apex and travels down again it will cross the x-axis with a velocity of `-DY`. If `DY` starts at `1-MinY` it's next step below the x-axis will be at the end of the target range.